text/0000-run-to-completion-async-fn.md

```diff
@@ -0,0 +1,908 @@
+internal mechanism (e.g. task-local storage) from the timer to the read call.
+However it could also be explicitly passed.
+
+This example also demonstrates the need for `#[completion] async` blocks.
+Without those, we could not pass the actual read method to `runtime::with_timeout`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+As described earlier, an `#[completion] async fn` would be transformed into a
+`RunToCompletionFuture`.
+
+## async transformation
+
+This section describes the necessary changes for the compiler in order to
+support the proposed async run to completion functions:
+
+### `async fn` -> `async` block transformation
+
+The compiler will need to desugar a function with a signature of
+
+```rust
+#[completion] async fn example(Args) -> Res {
+    body
+}
+```
+
+into
+
+```rust
+fn example(Args) -> RunToCompletionFuture<Output=Res> {
+    #[completion] async move {
+        body
+    }
+}
+```
+
+This requires the compiler to forward the #[completion] attribute to the
+generated async block.
+
+### `async` block -> generator transformation
+
+The translation of an async block into a generator/state-machine is expected to
+be mostly identical to the translation for existing async blocks.
+The only difference should be generating a different return type.
+
+### Type checks for `async` blocks
+
+Besides this the compiler will need to apply a slightly modified type checking
+behavior for `#[completion] async` blocks compared to `await` blocks:
+- Inside `#[completion] async` blocks users are allowed to `.await`
+  `RunToCompletionFuture`s and `Future`s.
+- Inside `async` blocks users are only allowed to `.await` `Future`s.
+
+## `RunToCompletionFuture` type definition
+
+This RFC proposes to keep `RunToCompletionFuture` closely aligned to the current
+`Future` type by defining it in the following fashion:
+
+```rust
+pub trait RunToCompletionFuture {
+    /// The type of value produced on completion.
+    type Output;
+
+    /// Attempt to resolve the future to a final value, registering
+    /// the current task for wakeup if the value is not yet available.
+    ///
+    /// # Return value
+    ///
+    /// This function returns:
+    ///
+    /// - [`Poll::Pending`] if the future is not ready yet
+    /// - [`Poll::Ready(val)`] with the result `val` of this future if it
+    ///   finished successfully.
+    ///
+    /// If a call to `poll()` call was issued to a `RunToCompletionFuture` that
+    /// returned `Pending`, the caller **must** call `poll()` again later,
+    /// until it returns `Ready`.
+    ///
+    /// Callers are not allowed `drop()` a future which returned `Pending` as
+    /// its last poll result. Futures are only allowed to be dropped if they
+    /// either had never been polled, or if the last `poll()` call returned `Ready`.
+    unsafe fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output>;
```

Comment:
If unfortunately seems like the only boundary that the language offers us at this point in time. The issue is that if someone does not poll the function to completion, but the content of the function expects that guarantee - then we would have a memory safety issue.

The example where a sub-task borrows something on the stack of the supertask, but doesn't end within the lifetime of the supertask and gets later resumed is such an example.

Reply:
Thanks for the insights. Our async Rust tasks can run in parallel on multiple threads though, which is quite a different beast.

> The issue is that if someone does not poll the function to completion, but the content of the function expects that guarantee - then we would have a memory safety issue.

Is this implication always true? I buy that the consistency and result of the function is inconsistent and "poisoned", but memory safety sounds like it would only apply in certain cases, such as the subtask-example.

Reply:
I'm not sure whether there is a big difference between requiring the callers to call `poll_drop` after  a panic, or simply continue to call `poll`. Since everything that the compiler puts into `poll_drop` could be in the same way in `poll`.

Maybe one of the arguments for a dedicated `poll_drop` is that it doesn't have the output value that `poll` has.
I think I would be ok with adding a `poll_drop` to the new `Future` - but more for pure compiler-generated drop glue than for humans to implement.

> Premature drop calls abort: leaking a polled RunToCompletionFuture would still be undefined behavior, but dropping it would call abort instead.

Yeah, that's the "simple way out". For my use-cases I would be rather fine with it, since I think panics + coroutines/futures is a super indeterministic thing anyway. But if others rely on the ability to handle panics in async run to completion contexts, it should indeed be taken into account.

> Is this implication always true? I buy that the consistency and result of the function is inconsistent and "poisoned", but memory safety sounds like it would only apply in certain cases, such as the subtask-example.

It doesn't matter whether its 1 example or thousands. If it's more than 0 examples it has to be labeled as `unsafe` in Rust. People rely on safe functions really being safe - even under heavy API misuse. That's why there also had been very unpleasant discussions for authors of crates which provided support for completion based IO. Those had APIs which very basically always safe, as long as one doesn't do really weird things with them. But still they got push-back from the community about not marking them `unsafe` because theoretically those could lead to memory issues.